chain/client/src/test_utils/test_env.rs

```diff
@@ -270,87 +274,92 @@ impl TestEnv {
-                            .process_chunk_state_witness(
-                                chunk_state_witness.clone(),
-                                PeerId::new(sender_public_key),
-                            )
+        let mut output = StateWitnessPropagationOutput {
+            found_differing_post_state_root_due_to_state_transitions: false,
+            chunk_hash_to_account_ids: HashMap::new(),
+        };
+        let network_adapters = self.network_adapters.clone();
+        for network_adapter in network_adapters {
+            network_adapter.handle_filtered(|request| match request {
+                PeerManagerMessageRequest::NetworkRequests(NetworkRequests::ChunkStateWitness(
+                    account_ids,
+                    state_witness,
+                )) => {
+                    // Process chunk state witness for each client.
+                    for account_id in account_ids.iter() {
+                        self.client(account_id)
+                            .process_chunk_state_witness(state_witness.clone(), PeerId::random())
                             .unwrap();
                     }
+
+                    // Update output.
+                    output.found_differing_post_state_root_due_to_state_transitions |=
+                        Self::found_differing_post_state_root_due_to_state_transitions(
+                            &state_witness,
+                        );
+
+                    output.chunk_hash_to_account_ids.insert(
+                        state_witness.inner.chunk_header.chunk_hash(),
+                        account_ids.into_iter().collect(),
+                    );
                     None
-                } else {
-                    Some(msg)
                 }
+                _ => Some(request),
             });
         }
-        StateWitnessPropagationOutput { found_differing_post_state_root_due_to_state_transitions }
+        output
     }
 
-    /// Waits for `CHUNK_ENDORSEMENTS_TIMEOUT` to receive at least one chunk
-    /// endorsement for the given chunk hashes.
+    /// Waits for `CHUNK_ENDORSEMENTS_TIMEOUT` to receive chunk endorsement for the given chunk hashes.
     /// Panics if it doesn't happen.
-    /// `chunk_hashes` maps hashes to pair of height at which chunk is included
-    /// and shard id for better debugging.
-    pub fn wait_for_chunk_endorsements(
+    /// `chunk_hash_to_account_ids` maps hashes to the set of account ids that are expected to endorse the chunk.
+    /// Note that we need to wait here as the chunk state witness is processed asynchronously.
+    pub fn wait_to_propagate_chunk_endorsements(
         &mut self,
-        mut chunk_hashes: HashMap<ChunkHash, (BlockHeight, ShardId)>,
+        chunk_hash_to_account_ids: &mut HashMap<ChunkHash, HashSet<AccountId>>,
     ) {
-        let _span = tracing::debug_span!(target: "test", "get_all_chunk_endorsements").entered();
+        let network_adapters = self.network_adapters.clone();
```

Comment:
Do we actually need to clone this?

Reply:
Yes as otherwise we would hold a mut reference to self within the loop. rustc wasn't happy